Fix wrong DOM order on suspend inside Fragment

jviide opened the pull request:
This pull request fixes issue #2106 and refactors the Suspense implementation for smaller code size.
 * Modify the Suspense lifecycle to be more explicit:
   * On the first suspended child set the `_detachOnNextRender` flag and force a new render with `setState`.
   * On the next render detach the old suspended children from DOM and replace them with an empty list of children.
 * Use a counter to keep track of the unresolved promises instead of an array of the promises.
 * Add a test for issue #2106.
 * Fix Suspense test "should correctly render nested Suspense components", it didn't return a promise.
 * Add a test for children suspending with the same promise.
 * Add a test ensuring that no children will be rendered if any of them suspends.

The changes in this PR bring down the compat.js.gz bundle size by 38 bytes (from 2700 B to 2662 B).

Fixes #2106.

ForsakenHarmony commented on compat/src/suspense.js:
```diff
@@ -30,24 +24,20 @@ options._catchError = function(error, newVNode, oldVNode) {
 export function Suspense(props) {
 	// we do not call super here to golf some bytes...
-	this._suspensions = [];
-	this._fallback = props.fallback;
```
was there some reason for storing this on the class?

jviide replied:
I'm not quite sure, but it might have had something to do how Suspense was implemented before this PR.

ForsakenHarmony replied:
https://github.com/preactjs/preact/commit/17f65e4dfd209d643117b0fab507c43eacdb4b96

Yeah probably some "legacy things"™

sventschui replied:
The fallback is later cloned using `cloneElement` to not re-use the vnode as this is not supported in our current code base. See line 87 prior to this PR

sventschui replied:
I guess the cloning is no longer needed as the fallback is now rendered as a child of a Fragment that is then cloned in coerceToVNode anyways if it was already used.

jviide replied:
Yeah I suppose so.

andrewiggins replied:
I think the problem was that the old implementation was manually calling `unmount`. So we had to make sure that the VNode that was called with `unmount` was the most up to date VNode. If we re-used a VNode (e.g. if a component re-suspends), `toChildArray` notices that this VNode has been used before (e.g. it has a `_component` property) and clones it. In the old implementation this was a problem because it mean that the reference to the `props.fallback` VNode was no longer in sync with the VNode that the diff was using.

To fix this, we cloned the `fallback` before going into the diff to skip the `toChildArray` clone, guaranteeing our reference always matched the one the diff was using.

Since this PR changes suspense to no longer manually call `unmount` we longer need to worry about VNode references, since we rely on the natural thing that `diff` does!

(This is from my memory of the previous PR, didn't re-verify it so I hope my memory isn't too off base lol)

marvinhagemeister commented on compat/src/suspense.js:
```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
+
+	return [
+		createElement(Fragment, null, state._suspended ? null : props.children),
+		createElement(Fragment, null, state._suspended && props.fallback)
```
I'm wondering if we can skip the inner `Fragments` here. Returning an array from `render` will create one and maybe we could just pass the suspensions as direct children.

sventschui replied:
Not sure whether the trick is to have two never un-mounting children rendered by suspense tricking our diff algo, isn't it?

jviide replied:
Yes. This was the smallest variant I found that keeps the tree consistent enough for the diff algorithm.

jviide replied:
Oh right, and I did test this with @marvinhagemeister's suggestion but the DOM elements started moving around again :)

marvinhagemeister commented on compat/test/browser/suspense.test.js:
```diff
@@ -916,7 +916,7 @@ describe('suspense', () => {
 		expect(scratch.innerHTML).to.equal(loadingHtml);
 
-		resolve1(() => <b>1</b>)
+		return resolve1(() => <b>1</b>)
```
👍

sventschui commented on compat/src/suspense.js:
```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
-		c.setState({ _parkedChildren: c._vnode._children });
-		detachDom(c._vnode._children);
-		c._vnode._children = [];
+	if (!c._suspensions++) {
```
I somewhen did an effort to de-duplicate `c._suspensions` when throwing the same promise multiple times which lead to duplicates in the `_suspensions` array and therefore to never un-suspend. 

Honestly I do not remember where this effort went as it seems to not be present in the current implementation.

We should check that this refactor still works when throwing the same promise multiple times.

jviide replied:
Yep, that's a good test. I'll add it to the suite.

I _suspect_ that the original array-based implementation (and this modified counter-based implementation) will survive the scenario even without special dedup logic. If the same promise is thrown from two children then it's added to the array twice (or the counter is incremented twice in the new implementation). When the promise resolves it's removed from the array twice (or the counter in decremented twice in the new implementation), which keeps the internal state consistent.

jviide replied:
I added the test, and it seems that both the original implementation and this modified one survive it just fine 👍

prateekbh commented on compat/src/suspense.js:
```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
+		this._detachOnNextRender = false;
+	}
+
+	return [
```
Any way we can just do a simple conditional based rendering here?

```js
return state._suspended?createElement(Fragment, null, props.fallback):createElement(Fragment, null, props.children)`
```

prateekbh replied:
just makes it much more readable and less magic

jviide replied:
It appears that Preact's diffing algorithm loses track with this variation as well :/

prateekbh replied:
:(

prateekbh replied:
also 

```js
return createElement(Fragment, null, state._suspended?props.children:props.fallback)
```

sorry, just trying to see if it can look less magical

jviide replied:
That variation fails the tests as well.

andrewiggins commented on compat/src/suspense.js:
```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
+	}
+
+	return [
+		createElement(Fragment, null, state._suspended ? null : props.children),
```
Double checking, this doesn't call the unmount lifecycles for children when a component suspends? Probably need to double check that we have a test that covers this.

sventschui replied:
That's actually a very valid point. Pretty sure this will call the unmount lifecycles what is different to react and not really the meaning of suspense

jviide replied:
There appears to be tests for such scenarios already, which is awesome:
 * [`should not call lifecycle methods when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L230-L283)
 * [`should call fallback's lifecycle methods when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L285-L339)
 * [`should keep state of children when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L341-L388)

The first one (`should not call lifecycle methods when suspending`) tests that the lifecycle hooks of a component with a sibling that suspends. I added a new test that tests the lifecycle hooks of a component that itself suspends. I assumed that the `componentDidMount` shouldn't be called for such a component when it suspends, as React doesn't seem to do so either (see this [Codesandbox example](https://codesandbox.io/s/componentdidmount-not-called-for-suspended-component-wvh66)).

sventschui replied:
But how does that first test succeed? I‘d expect the lifecycles being called with this implementation. Will try to free some time to understand this better 🤔

sventschui replied:
Ah now I see that‘s the detachedClone fiddling you do in render 👍👌

developit commented on compat/src/suspense.js:
```diff
@@ -1,10 +1,4 @@
-import {
-	Component,
-	createElement,
-	_unmount as unmount,
```
this is a nice win.

jviide replied:
`compat/src/index.js` imports and uses `unmount`, namely the `Portal` implementation.

jviide commented:
Updated the implementation. Now DOM is not modified manually at all. Instead the suspended component is removed/detached/lifted from the rendered tree to keep its state intact, and added back to the rendered tree when the suspension gets resolved.

The `[<Fragment>{props.children?}</Fragment>, props.fallback?]` structure that `render` returns is there because it makes bookkeeping easier: We always know that the (potentially suspended) children are inside the first child of the Suspense component.

This brings the compat.js.gz bundle size further down by 7 bytes (now 2655 B vs. the 2700 B in the master branch, 45 B reduction in total).

marvinhagemeister commented:
Merging, coveralls is acting up strangely again...

sventschui commented:
maybe coveralls is right. There are some missed branches in the suspense impl. Might be worth looking at these. I might find some time to do so later today

andrewiggins commented:
Note, if you click on the "coverage" badge you can see exactly which lines and branches aren't covered

marvinhagemeister commented:
There were no coverage changes related to any files in this PR. The one were coverage went down was the entry file of `preact/debug` :man_shrugging:

prateekbh commented:
One problem: Anything relying on `props.children` now gets additional nodes due to the two children.

Anything traversing props.children will probably break. Unless thats an anti pattern for libraries that I am not aware of

jviide commented on debug/test/browser/debug-suspense.test.js:
```diff
@@ -9,6 +9,14 @@ import {
 
 /** @jsx createElement */
 
+async function waitResolved(suspense) {
```
Sweet! Sorry that I didn't react to this in time. If this PR caused merge conflicts to your branch I can create a patch to resolve them.

jviide commented:
@prateekbh `props.children` do not get modified by this code. Could you give an example of a case that you suspect might break?

prateekbh commented:
You're right I mistakenly added a ghost space in my tests and thought it came because of this

prateekbh commented:
All good 👍👍👍
